### PR TITLE
Enforce that `<amp-ad>` is not in fixed position containers.

### LINKED ISSUE #690: Disallow `<amp-ad>` within position:fixed elements
This should be a runtime check, runtime error and deny rendering. Docs need to be updated.

This may be relaxed in the future, but we do not want full page over lays on AMP pages, because users don't like them.

### PR DESCRIPTION
**Pull request by cramforce**
Closes #690

### CODE REVIEW

**Comment by erwinmombay** [comment]
can you add this to the diagram?

File: builtins/amp-ad.js [hunk 1]
```diff
@@ -134,8 +135,34 @@ export function installAd(win) {
...
+     * anyway.
+     */
+    isPositionFixed() {
+      let el = this.element;
+      do {
+        if (this.win.getComputedStyle(el).position == 'fixed') {
+          return true;
+        }
+        el = el.parentNode;
+      } while (el.getAttribute || el == this.win.document.body);
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: `el != body`?

File: builtins/amp-ad.js [hunk 2]
```diff
@@ -134,8 +135,34 @@ export function installAd(win) {
...
 
+    /**
+     * @override
+     */
+    onLayoutMeasure() {
+      this.isInFixedContainer_ = this.isPositionFixed();
+    }
+
+    /**
+     * @return {boolean} whether this element or its ancestors have position
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: Does it mean that `amp-ad` is not allowed in the lightbox?

File: builtins/amp-ad.md [hunk 3]
```diff
@@ -38,6 +38,12 @@ resources in AMP. It requires a `type` argument that select what ad network is d
     data-aax_src="302">
 </amp-ad>
 ```
+
+#### Styling
+
+`<amp-ad>` elements may not themselves have or be placed in containers that have CSS `position: fixed` set.
```
Comment (dvoytenko) [inline builtins/amp-ad.md]: So, one possible case here is the desktop layout where ads are pushed to the right? Are those bad or will we leave them for later?

File: src/base-element.js [hunk 4]
```diff
@@ -497,4 +500,11 @@ export class BaseElement {
   cancelFullOverlay() {
     viewerFor(this.element.ownerDocument.defaultView).cancelFullOverlay();
```
Comment (dvoytenko) [inline src/base-element.js]: s/this.element.ownerDocument.defaultView/this.win

File: src/custom-element.js [hunk 5]
```diff
@@ -418,6 +418,7 @@ export function createAmpElementProto(win, name, implementationClass) {
     if (this.isUpgraded()) {
       this.implementation_.layoutWidth_ = this.layoutWidth_;
     }
+    this.implementation_.onLayoutMeasure();
```
Comment (dvoytenko) [inline src/custom-element.js]: Should only do it when upgraded. E.g. in the `if` above. But we will also need to consider how to set it after upgrade.

File: builtins/amp-ad.js (hunk 1 above)
Comment (jridgewell) [inline builtins/amp-ad.js]: You'd need an `&&`, too.

File: src/base-element.js (hunk 4 above)
Comment (cramforce) [inline src/base-element.js]: Done

File: builtins/amp-ad.md (hunk 3 above)
Comment (cramforce) [inline builtins/amp-ad.md]: Will leave for later

File: src/custom-element.js (hunk 5 above)
Comment (cramforce) [inline src/custom-element.js]: I actually think the right thing here would be to schedule it for later call, but I added a TODO for now.

File: builtins/amp-ad.js (hunk 2 above)
Comment (cramforce) [inline builtins/amp-ad.js]: Added POSITION_FIXED_TAG_WHITELIST

File: builtins/amp-ad.js (hunk 1 above)
Comment (cramforce) [inline builtins/amp-ad.js]: Done. Duh.

File: builtins/amp-ad.js [hunk 6]
```diff
@@ -134,8 +140,38 @@ export function installAd(win) {
...
+     * fixed (unless they are POSITION_FIXED_TAG_WHITELIST).
+     * This should only be called when a layout on the page was just forced
+     * anyway.
+     */
+    isPositionFixed() {
+      let el = this.element;
+      let body = el.ownerDocument.body;
+      do {
+        if (!POSITION_FIXED_TAG_WHITELIST[el.tagName] &&
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: You probably want to exit as soon as you hit a POSITION_FIXED_TAG_WHITELIST parent.

File: src/custom-element.js [hunk 7]
```diff
@@ -418,6 +418,8 @@ export function createAmpElementProto(win, name, implementationClass) {
     if (this.isUpgraded()) {
       this.implementation_.layoutWidth_ = this.layoutWidth_;
     }
+    // TODO(malteubl): Forward for stubbed elements.
+    this.implementation_.onLayoutMeasure();
```
Comment (dvoytenko) [inline src/custom-element.js]: So, right now it will work only because amp-ad is builtin, right?

**Comment by dvoytenko** [comment]
LGTM with a couple of comments.

File: builtins/amp-ad.js (hunk 6 above)
Comment (cramforce) [inline builtins/amp-ad.js]: Done

File: src/custom-element.js (hunk 7 above)
Comment (cramforce) [inline src/custom-element.js]: Yes. Added docs with that respect.

**Pull request merged**